src/panels/lovelace/cards/hui-sensor-card.ts

```diff
@@ -0,0 +1,375 @@
+      } else {
+        key = Math.floor(key);
+      }
+      if (!res[key]) {
+        res[key] = [];
+      }
+      res[key].push(item);
+      return res;
+    };
+    history = history.reduce((res, item) => reduce(res, item), []);
+    if (detail > 1) {
+      history = history.map((entry) =>
+        entry.reduce((res, item) => reduce(res, item, true), [])
+      );
+    }
+    return this._calcPoints(history, hours, width, detail, min, max);
+  }
+
+  private _calcPoints(
+    history,
+    hours,
+    width: number,
+    detail: number,
+    min: number,
+    max: number
+  ): number[][] {
+    const coords = [] as number[][];
+    const margin = 5;
+    const height = 80;
+    width -= 10;
+    let yRatio = (max - min) / height;
+    yRatio = yRatio !== 0 ? yRatio : height;
+    let xRatio = width / (hours - (detail === 1 ? 1 : 0));
+    xRatio = isFinite(xRatio) ? xRatio : width;
+    const getCoords = (item, i, offset = 0, depth = 1) => {
+      if (depth > 1) {
+        return item.forEach((subItem, index) =>
+          getCoords(subItem, i, index, depth - 1)
+        );
+      }
+      const average =
+        item.reduce((sum, entry) => sum + parseFloat(entry.state), 0) /
+        item.length;
+
+      const x = xRatio * (i + offset / 6) + margin;
+      const y = height - (average - min) / yRatio + margin * 2;
+      return coords.push([x, y]);
+    };
+
+    history.forEach((item, i) => getCoords(item, i, 0, detail));
+    if (coords.length === 1) {
+      coords[1] = [width + margin, coords[0][1]];
+    }
+    coords.push([width + margin, coords[coords.length - 1][1]]);
+    return coords;
+  }
+
+  private _getPath(coords: number[][]): string {
+    let next;
+    let Z;
+    const X = 0;
+    const Y = 1;
+    let path = "";
+    let last = coords.filter(Boolean)[0];
+
+    path += `M ${last[X]},${last[Y]}`;
+
+    for (const coord of coords) {
+      next = coord;
+      Z = this._midPoint(last[X], last[Y], next[X], next[Y]);
+      path += ` ${Z[X]},${Z[Y]}`;
+      path += ` Q${next[X]},${next[Y]}`;
+      last = next;
+    }
+
+    path += ` ${next[X]},${next[Y]}`;
+    return path;
+  }
+
+  private _midPoint(
+    _Ax: number,
+    _Ay: number,
+    _Bx: number,
+    _By: number
+  ): number[] {
+    const _Zx = (_Ax - _Bx) / 2 + _Bx;
+    const _Zy = (_Ay - _By) / 2 + _By;
+    return [_Zx, _Zy];
+  }
+
+  private async _getHistory(): Promise<string> {
+    const endTime = new Date();
+    const startTime = new Date();
+    startTime.setHours(endTime.getHours() - this._config!.hours_to_show!);
+
+    const stateHistory = await this._fetchRecent(startTime, endTime);
+
+    if (stateHistory[0].length < 1) {
+      return "";
+    }
+    const coords = this._coordinates(
+      stateHistory[0],
+      this._config!.hours_to_show!,
+      500,
+      this._config!.detail!
+    );
+    return this._getPath(coords);
+  }
+
+  private async _fetchRecent(startTime: Date, endTime: Date): Promise<{}> {
```

Comment:
We should move this to `src/data/history.ts`

Reply:
Looks like it was already there